lucille-core/src/main/java/com/kmwllc/lucille/stage/util/Py4JRuntime.java

```diff
@@ -0,0 +1,422 @@
+      log.info("No requirementsPath specified; skipping requirements installation.");
+      return;
+    }
+    log.info("Installing Python requirements from {} using venv python {}", requirementsPath, venvPythonPath);
+    try {
+      Process installProc = new ProcessBuilder(
+          venvPythonPath, "-m", "pip", "install", "-r", requirementsPath)
+          .redirectErrorStream(true)
+          .start();
+      StringBuilder installOutput = new StringBuilder();
+      try (java.io.BufferedReader installReader = new java.io.BufferedReader(
+          new java.io.InputStreamReader(installProc.getInputStream()))) {
+        String line;
+        while ((line = installReader.readLine()) != null) {
+          installOutput.append(line).append("\n");
+        }
+      }
+      int installExit = installProc.waitFor();
+      log.info("pip install -r output:\n{}", installOutput.toString().trim());
+      if (installExit != 0) {
+        throw new StageException("Failed to install requirements from " + requirementsPath + ". Output: " + installOutput);
+      }
+    } catch (Exception e) {
+      throw new StageException("Failed to install requirements from " + requirementsPath + ": " + e.getMessage(), e);
+    }
+  }
+
+  private void logPythonEnvironment() throws StageException {
+    try {
+      // Get the Python executable path using venv python
+      Process execPathProc = new ProcessBuilder(venvPythonPath, "-c", "import sys; print(sys.executable)")
+          .redirectErrorStream(true).start();
+      StringBuilder execPathOutput = new StringBuilder();
+      try (java.io.BufferedReader execPathReader = new java.io.BufferedReader(
+          new java.io.InputStreamReader(execPathProc.getInputStream()))) {
+        String line;
+        while ((line = execPathReader.readLine()) != null) {
+          execPathOutput.append(line).append("\n");
+        }
+      }
+      execPathProc.waitFor();
+      log.info("Python executable path (from sys.executable):\n{}", execPathOutput.toString().trim());
+
+      // Log all sys.path entries
+      Process sysPathProc = new ProcessBuilder(venvPythonPath, "-u", "-c", "import sys; [print(p) for p in sys.path]")
+          .redirectErrorStream(true).start();
+      StringBuilder sysPaths = new StringBuilder();
+      try (java.io.BufferedReader sysPathReader = new java.io.BufferedReader(
+          new java.io.InputStreamReader(sysPathProc.getInputStream()))) {
+        String line;
+        while ((line = sysPathReader.readLine()) != null) {
+          sysPaths.append(line).append("\n");
+        }
+      }
+      sysPathProc.waitFor();
+      log.info("Python sys.path entries:\n{}", sysPaths.toString().trim());
+    } catch (Exception e) {
+      log.error("Error logging Python environment", e);
+      throw new StageException("Failed to log Python environment: " + e.getMessage(), e);
+    }
+  }
+
+  private static int allocatePort(int configured) throws StageException {
```

Comment:
suggest just making this `private static synchronized` and removing the internal sychronized blocks for simplicity. Same with releasePort(). Can also remove `private static final Object portLock`
These methods are only called once upon stage startup so we do not need to optimize them for concurrency